lib/srv/desktop/rdp/rdpclient/src/rdpdr/mod.rs

```diff
@@ -2739,6 +3002,152 @@ impl DeviceReadResponse {
+
+        Ok(Self {
+            device_io_request,
+            length,
+            offset,
+            write_data,
+        })
+    }
+}
+
+/// 2.2.1.5.4 Device Write Response (DR_WRITE_RSP)
+/// https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/58160a47-2379-4c4a-a99d-24a1a666c02a
+#[derive(Debug)]
+pub struct DeviceWriteResponse {
+    /// The CompletionId field of this header MUST match a Device I/O Request (section 2.2.1.4) message that had the MajorFunction field set to IRP_MJ_WRITE.
+    device_io_reply: DeviceIoResponse,
+    /// Number of bytes written in response to the write request.
+    length: u32,
+}
+
+impl DeviceWriteResponse {
+    fn new(device_io_request: &DeviceIoRequest, io_status: NTSTATUS, length: u32) -> Self {
+        Self {
+            device_io_reply: DeviceIoResponse::new(
+                device_io_request,
+                NTSTATUS::to_u32(&io_status).unwrap(),
+            ),
+            length,
+        }
+    }
+
+    fn encode(&self) -> RdpResult<Vec<u8>> {
+        let mut w = vec![];
+        w.extend_from_slice(&self.device_io_reply.encode()?);
+        w.write_u32::<LittleEndian>(self.length)?;
+        // 1 byte padding
+        w.write_u32::<LittleEndian>(0)?;
+        Ok(w)
+    }
+}
+
+/// 2.2.3.4.9 Client Drive Set Information Response (DR_DRIVE_SET_INFORMATION_RSP)
+/// https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/16b893d5-5d8b-49d1-8dcb-ee21e7612970
+#[derive(Debug)]
+#[allow(dead_code)]
+struct ClientDriveSetInformationResponse {
+    device_io_reply: DeviceIoResponse,
+    /// This field MUST be equal to the Length field in the Server Drive Set Information Request (section 2.2.3.3.9).
+    length: u32,
+}
+
+impl ClientDriveSetInformationResponse {
+    fn new(device_read_request: &ServerDriveSetInformationRequest, io_status: NTSTATUS) -> Self {
+        Self {
+            device_io_reply: DeviceIoResponse::new(
+                &device_read_request.device_io_request,
+                NTSTATUS::to_u32(&io_status).unwrap(),
+            ),
+            length: device_read_request.set_buffer.len() as u32,
+        }
+    }
+
+    fn encode(&self) -> RdpResult<Vec<u8>> {
+        let mut w = vec![];
+        w.extend_from_slice(&self.device_io_reply.encode()?);
+        w.write_u32::<LittleEndian>(self.length)?;
+        // 1 byte padding
+        w.write_u32::<LittleEndian>(0)?;
+        Ok(w)
+    }
+}
+
+/// 2.2.3.3.9 Server Drive Set Information Request (DR_DRIVE_SET_INFORMATION_REQ)
+/// https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/b5d3104b-0e42-4cf8-9059-e9fe86615e5c
+#[derive(Debug)]
+#[allow(dead_code)]
+struct ServerDriveSetInformationRequest {
+    /// The MajorFunction field in the DR_DEVICE_IOREQUEST header MUST be set to IRP_MJ_SET_INFORMATION.
+    device_io_request: DeviceIoRequest,
+    file_information_class_level: FileInformationClassLevel,
+    /// TODO(LKozlowski): change to enum?
```

Comment:
I think we want to make it a
https://github.com/gravitational/teleport/blob/09e253983cceaaf51e72ffe6360e07a30b59620b/lib/srv/desktop/rdp/rdpclient/src/rdpdr/mod.rs#L2191-L2197

and add a `decode` to all of the implementations which we can call in `ServerDriveSetInformationRequest::decode`. 

(See how I've implemented `encode`)
https://github.com/gravitational/teleport/blob/09e253983cceaaf51e72ffe6360e07a30b59620b/lib/srv/desktop/rdp/rdpclient/src/rdpdr/mod.rs#L2200-L2210

Reply:
I was thinking about it and that's why I wrote a comment: `change to enum?`😄  I'll implement it here so we don't have to do it in the other PRs